### DISCUSSION TITLE
Some `aria-` variants out of the box?

### DISCUSSION THREAD

**Discussion by MichaelAllenWarner** [tailwindlabs/tailwindcss#5659] [Ideas]
For interactive components that require JavaScript for accessibility, it's often convenient (and arguably good practice) to use `aria-` attributes as selectors for styling purposes (e.g., `[aria-hidden="true"]`, `[aria-selected="true"]`, `[aria-expanded="true"]`). Might be worth including variants for a few of these out of the box. Doing so might have the added benefit of encouraging some developers to think more about accessibility.

**Comment by MerlinMason** [comment]
Yeah I think being able to specify arbitrary variants would be useful, would also simplify how libraries like Radix or even Tailwinds HeadlessUI could style state.

For example from the HeadlessUI docs:
```js
 <RadioGroup value={plan} onChange={setPlan}>
      <RadioGroup.Label>Plan</RadioGroup.Label>

      <RadioGroup.Option value="startup">
        {({ checked }) => (
          <span className={checked ? 'bg-blue-200' : ''}>Startup</span>
        )}
      </RadioGroup.Option>
      <RadioGroup.Option value="business">
        {({ checked }) => (
          <span className={checked ? 'bg-blue-200' : ''}>Business</span>
        )}
      </RadioGroup.Option>
      <RadioGroup.Option value="enterprise">
        {({ checked }) => (
          <span className={checked ? 'bg-blue-200' : ''}>Enterprise</span>
        )}
      </RadioGroup.Option>
</RadioGroup>
```

Could be simplified to something like:
```js
 <RadioGroup value={plan} onChange={setPlan}>
      <RadioGroup.Label>Plan</RadioGroup.Label>

      <RadioGroup.Option value="startup" className="[aria-checked=true]:bg-blue-200">Startup</RadioGroup.Option>
      <RadioGroup.Option value="business" className="[aria-checked=true]:bg-blue-200">Business</RadioGroup.Option>
      <RadioGroup.Option value="enterprise" className="[aria-checked=true]:bg-blue-200">Enterprise</RadioGroup.Option>
</RadioGroup>
```

**Reply by MichaelAllenWarner** [reply]
I actually started a separate discussion for custom selectors: https://github.com/tailwindlabs/tailwindcss/discussions/5473

**Reply by thibaudcolas** [reply]
Excellent example

**Reply by lukeocodes** [reply]
Literally exactly what I am trying to do, given as an example. Nice :) Hope this makes it's way in.

**Reply by MichaelAllenWarner** [reply]
@lukeocodes It already did! See https://tailwindcss.com/docs/adding-custom-styles#arbitrary-variants

**Reply by wongjn** [reply]
There is [arbitrary variants](https://tailwindcss.com/docs/adding-custom-styles#arbitrary-variants) and [some `aria-*` variants](https://tailwindcss.com/docs/hover-focus-and-other-states#aria-states) baked in.

**Reply by lukeocodes** [reply]
I was looking at arbitrary varients and I completely scrolled over that. Thanks both 😅

**Comment by sntran** [comment]
I think this is a great idea. Right now Tailwind has built-in variant for `open`, which matches on `[open]` internally.

If arbitrary attribute/selector is supported, then we can do `[open]` directly.

**Comment by thibaudcolas** [comment]
Looks like there is a [tailwindcss-aria-attributes](https://github.com/thoughtbot/tailwindcss-aria-attributes) plugin which does this, but yes as others mention this really should be built-in like `[open]`. The big one in my opinion is `aria-expanded:`, which is used all the time for [disclosure widgets](https://w3c.github.io/aria-practices/#disclosure).

Code written like this is much more readable than the equivalent `{expanded ? 'bg-blue-200' : ''}`, both in the source, and the HTML output, since the classes are stable regardless of the component’s state.
